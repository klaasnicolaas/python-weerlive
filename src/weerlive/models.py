"""Asynchronous Python client for Weerlive."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, time

from mashumaro import field_options
from mashumaro.config import BaseConfig
from mashumaro.mixins.orjson import DataClassORJSONMixin
from mashumaro.types import SerializationStrategy


class TimeStrategy(SerializationStrategy):
    """String serialization strategy to handle the time format."""

    def serialize(self, value: time) -> str:
        """Serialize time to string."""
        return value.strftime("%H:%M")

    def deserialize(self, value: str) -> time:
        """Deserialize string to time."""
        return datetime.strptime(value, "%H:%M").astimezone().time()


@dataclass
# pylint: disable-next=too-many-instance-attributes
class Weather(DataClassORJSONMixin):
    """Object representing an Weather model.

    Attributes
    ----------
        location: The location of the weather.
        temperature: Current temperature.
        wind_chill: Perceived temperature.
        summary: Description of weather conditions.
        humidity: Relative air humidity.
        wind_d: Wind direction.
        wind_ms: Wind speed in ms.
        wind_f: Wind force (Beaufort).
        wind_kn: Wind speed in knots.
        wind_kmh: Wind speed in km/h.
        air_pressure: Air pressure in hPa.
        air_pressure_mmhg: Air pressure in mmHg.
        dew_point: Dew point in degrees Celsius.
        visibility: Visibility in kilometers.
        forecast: Weather forecast for the next 24 hours.
        sun_up: Time of sunrise.
        sun_down: Time of sunset.
        icon: Weather icon.

        d0_weather: Weather forecast for today.
        d0_temp_max: Maximum temperature for today.
        d0_temp_min: Minimum temperature for today.
        d0_wind_f: Wind force (Beaufort) for today.
        d0_wind_kn: Wind speed in knots for today.
        d0_wind_ms: Wind speed in ms for today.
        d0_wind_kmh: Wind speed in km/h for today.
        d0_wind_d: Wind direction for today.
        d0_wind_ddg: Wind direction in degrees for today.
        d0_rainfall: Change of rainfall in % for today.
        d0_sun: Chance of sunshine in % for today.

        d1_weather: Weather forecast for tomorrow.
        d1_temp_max: Maximum temperature for tomorrow.
        d1_temp_min: Minimum temperature for tomorrow.
        d1_wind_f: Wind force (Beaufort) for tomorrow.
        d1_wind_kn: Wind speed in knots for tomorrow.
        d1_wind_ms: Wind speed in ms for tomorrow.
        d1_wind_kmh: Wind speed in km/h for tomorrow.
        d1_wind_d: Wind direction for tomorrow.
        d1_wind_ddg: Wind direction in degrees
        d1_rainfall: Change of rainfall in % for tomorrow.
        d1_sun: Chance of sunshine in % for tomorrow.

        d2_weather: Weather forecast for the day after tomorrow.
        d2_temp_max: Maximum temperature for the day after tomorrow.
        d2_temp_min: Minimum temperature for the day after tomorrow.
        d2_wind_f: Wind force (Beaufort) for the day after tomorrow.
        d2_wind_kn: Wind speed in knots for the day after tomorrow.
        d2_wind_ms: Wind speed in ms for the day after tomorrow.
        d2_wind_kmh: Wind speed in km/h for the day after tomorrow.
        d2_wind_d: Wind direction for the day after tomorrow.
        d2_wind_ddg: Wind direction in degrees
        d2_rainfall: Change of rainfall in % for the day after tomorrow.
        d2_sun: Chance of sunshine in % for the day after tomorrow.
    """

    # pylint: disable-next=too-few-public-methods
    class Config(BaseConfig):
        """Mashumaro configuration."""

        serialization_strategy = {time: TimeStrategy()}  # noqa: RUF012
        serialize_by_alias = True

    location: str = field(metadata=field_options(alias="plaats"))
    temperature: float = field(metadata=field_options(alias="temp"))
    wind_chill: float = field(metadata=field_options(alias="gtemp"))
    summary: str = field(metadata=field_options(alias="samenv"))
    humidity: int = field(metadata=field_options(alias="lv"))
    wind_d: str = field(metadata=field_options(alias="windr"))
    wind_ms: float = field(metadata=field_options(alias="windms"))
    wind_f: int = field(metadata=field_options(alias="winds"))
    wind_kn: float = field(metadata=field_options(alias="windk"))
    wind_kmh: float = field(metadata=field_options(alias="windkmh"))
    air_pressure: float = field(metadata=field_options(alias="luchtd"))
    air_pressure_mmhg: int = field(metadata=field_options(alias="ldmmhg"))
    dew_point: float = field(metadata=field_options(alias="dauwp"))
    visibility: int = field(metadata=field_options(alias="zicht"))
    forecast: str = field(metadata=field_options(alias="verw"))
    sun_up: time = field(metadata=field_options(alias="sup"))
    sun_down: time = field(metadata=field_options(alias="sunder"))
    icon: str = field(metadata=field_options(alias="image"))

    d0_weather: str = field(metadata=field_options(alias="d0weer"))
    d0_temp_max: int = field(metadata=field_options(alias="d0tmax"))
    d0_temp_min: int = field(metadata=field_options(alias="d0tmin"))
    d0_wind_f: int = field(metadata=field_options(alias="d0windk"))
    d0_wind_kn: int = field(metadata=field_options(alias="d0windknp"))
    d0_wind_ms: int = field(metadata=field_options(alias="d0windms"))
    d0_wind_kmh: int = field(metadata=field_options(alias="d0windkmh"))
    d0_wind_d: str = field(metadata=field_options(alias="d0windr"))
    d0_wind_ddg: int = field(metadata=field_options(alias="d0windrgr"))
    d0_rainfall: int = field(metadata=field_options(alias="d0neerslag"))
    d0_sun: int = field(metadata=field_options(alias="d0zon"))

    d1_weather: str = field(metadata=field_options(alias="d1weer"))
    d1_temp_max: int = field(metadata=field_options(alias="d1tmax"))
    d1_temp_min: int = field(metadata=field_options(alias="d1tmin"))
    d1_wind_f: int = field(metadata=field_options(alias="d1windk"))
    d1_wind_kn: int = field(metadata=field_options(alias="d1windknp"))
    d1_wind_ms: int = field(metadata=field_options(alias="d1windms"))
    d1_wind_kmh: int = field(metadata=field_options(alias="d1windkmh"))
    d1_wind_d: str = field(metadata=field_options(alias="d1windr"))
    d1_wind_ddg: int = field(metadata=field_options(alias="d1windrgr"))
    d1_rainfall: int = field(metadata=field_options(alias="d1neerslag"))
    d1_sun: int = field(metadata=field_options(alias="d1zon"))

    d2_weather: str = field(metadata=field_options(alias="d2weer"))
    d2_temp_max: int = field(metadata=field_options(alias="d2tmax"))
    d2_temp_min: int = field(metadata=field_options(alias="d2tmin"))
    d2_wind_f: int = field(metadata=field_options(alias="d2windk"))
    d2_wind_kn: int = field(metadata=field_options(alias="d2windknp"))
    d2_wind_ms: int = field(metadata=field_options(alias="d2windms"))
    d2_wind_kmh: int = field(metadata=field_options(alias="d2windkmh"))
    d2_wind_d: str = field(metadata=field_options(alias="d2windr"))
    d2_wind_ddg: int = field(metadata=field_options(alias="d2windrgr"))
    d2_rainfall: int = field(metadata=field_options(alias="d2neerslag"))
    d2_sun: int = field(metadata=field_options(alias="d2zon"))
