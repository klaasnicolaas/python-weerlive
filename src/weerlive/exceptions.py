"""Asynchronous Python client for Weerlive."""


class WeerliveError(Exception):
    """General exception for Weerlive."""


class WeerliveConnectionError(WeerliveError):
    """Weerlive connection exception."""


class WeerliveAuthenticationError(WeerliveError):
    """Weerlive authentication exception."""


class WeerliveRateLimitError(WeerliveError):
    """Weerlive rate limit exception."""
