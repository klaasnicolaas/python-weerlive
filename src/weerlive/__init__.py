"""Asynchronous Python client for Weerlive."""

from .exceptions import (
    WeerliveAuthenticationError,
    WeerliveConnectionError,
    WeerliveError,
    WeerliveRateLimitError,
)
from .models import Weather
from .weerlive import Weerlive

__all__ = [
    "Weather",
    "Weerlive",
    "WeerliveAuthenticationError",
    "WeerliveConnectionError",
    "WeerliveError",
    "WeerliveRateLimitError",
]
