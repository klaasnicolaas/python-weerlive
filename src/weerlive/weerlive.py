"""Asynchronous Python client for Weerlive."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Mapping, Self, cast

from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import (
    WeerliveAuthenticationError,
    WeerliveConnectionError,
    WeerliveError,
    WeerliveRateLimitError,
)
from .models import Weather


@dataclass
class Weerlive:
    """Main class for handling connections with the Weerlive API."""

    api_key: str
    latitude: float
    longitude: float
    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: Mapping[str, Any] | None = None,
    ) -> Any:
        version = metadata.version(__package__)
        url = URL.build(scheme="https", host="weerlive.nl", path="/api/").join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonWeerlive/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        if self.api_key is None or self.api_key == "":
            msg = "No API key provided"
            raise WeerliveAuthenticationError(msg)

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Weerlive API."
            raise WeerliveConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with the Weerlive API."
            raise WeerliveConnectionError(msg) from exception

        # The API does not use status codes for error handling
        # Errors are therefore still returned with status 200
        response_text = await response.text()
        if "Vraag eerst een API-key op" in response_text:
            msg = "The given API key is invalid"
            raise WeerliveAuthenticationError(msg)
        if "Dagelijkse limiet" in response_text:
            msg = "The API rate limit has been exceeded"
            raise WeerliveRateLimitError(msg)

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            msg = "Unexpected content type response from the Weerlive API"
            raise WeerliveError(
                msg,
                {"Content-Type": content_type, "response": response_text},
            )

        return cast(dict[str, Any], await response.json())

    async def weather(self) -> Weather:
        """Get the current weather forecast.

        Returns
        -------
            A Weather data object from the API.
        """
        data = await self._request(
            "json-data-10min.php",
            params={
                "key": self.api_key,
                "locatie": f"{self.latitude},{self.longitude}",
            },
        )
        return Weather.from_dict(data["liveweer"][0])

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The Weerlive object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
