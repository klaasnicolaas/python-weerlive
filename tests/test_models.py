"""Test the models for Weerlive."""
from __future__ import annotations

from aiohttp import ClientSession
from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from weerlive import Weather, Weerlive

from . import load_fixtures


async def test_weather_data(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test weather data function."""
    aresponses.add(
        "weerlive.nl",
        "/api/json-data-10min.php",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("weather.json"),
        ),
    )
    async with ClientSession() as session:
        client = Weerlive(
            api_key="test",
            longitude=66.6699498,
            latitude=26.0317749,
            session=session,
        )
        weather: Weather = await client.weather()
        assert weather == snapshot
        assert weather.to_json() == snapshot
