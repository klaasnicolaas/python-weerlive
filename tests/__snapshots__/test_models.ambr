# serializer version: 1
# name: test_weather_data
  Weather(location='Amsterdam', temperature=9.1, wind_chill=4.8, summary='Regen', humidity=83, wind_d='ZW', wind_ms=11.0, wind_f=6, wind_kn=21.4, wind_kmh=39.6, air_pressure=996.0, air_pressure_mmhg=747, dew_point=6.0, visibility=5, forecast='Aan de kust zware windstoten. Perioden met regen', sun_up=datetime.time(8, 36), sun_down=datetime.time(16, 30), icon='regen', d0_weather='bewolkt', d0_temp_max=11, d0_temp_min=6, d0_wind_f=4, d0_wind_kn=14, d0_wind_ms=7, d0_wind_kmh=26, d0_wind_d='ZW', d0_wind_ddg=225, d0_rainfall=17, d0_sun=11, d1_weather='halfbewolkt_regen', d1_temp_max=10, d1_temp_min=8, d1_wind_f=3, d1_wind_kn=10, d1_wind_ms=5, d1_wind_kmh=19, d1_wind_d='W', d1_wind_ddg=270, d1_rainfall=70, d1_sun=30, d2_weather='regen', d2_temp_max=11, d2_temp_min=8, d2_wind_f=3, d2_wind_kn=10, d2_wind_ms=5, d2_wind_kmh=19, d2_wind_d='W', d2_wind_ddg=270, d2_rainfall=80, d2_sun=10)
# ---
# name: test_weather_data.1
  '{"plaats":"Amsterdam","temp":9.1,"gtemp":4.8,"samenv":"Regen","lv":83,"windr":"ZW","windms":11.0,"winds":6,"windk":21.4,"windkmh":39.6,"luchtd":996.0,"ldmmhg":747,"dauwp":6.0,"zicht":5,"verw":"Aan de kust zware windstoten. Perioden met regen","sup":"08:36","sunder":"16:30","image":"regen","d0weer":"bewolkt","d0tmax":11,"d0tmin":6,"d0windk":4,"d0windknp":14,"d0windms":7,"d0windkmh":26,"d0windr":"ZW","d0windrgr":225,"d0neerslag":17,"d0zon":11,"d1weer":"halfbewolkt_regen","d1tmax":10,"d1tmin":8,"d1windk":3,"d1windknp":10,"d1windms":5,"d1windkmh":19,"d1windr":"W","d1windrgr":270,"d1neerslag":70,"d1zon":30,"d2weer":"regen","d2tmax":11,"d2tmin":8,"d2windk":3,"d2windknp":10,"d2windms":5,"d2windkmh":19,"d2windr":"W","d2windrgr":270,"d2neerslag":80,"d2zon":10}'
# ---
